metronome/checkpointing/service/src/io/iohk/metronome/checkpointing/service/CheckpointingService.scala

```diff
@@ -0,0 +1,208 @@
+package io.iohk.metronome.checkpointing.service
+
+import cats.Applicative
+import cats.data.{NonEmptyList, NonEmptyVector, OptionT}
+import cats.effect.concurrent.{Ref, Semaphore}
+import cats.effect.{Concurrent, Resource, Sync}
+import cats.implicits._
+import io.iohk.metronome.checkpointing.CheckpointingAgreement
+import io.iohk.metronome.checkpointing.models.{
+  Block,
+  CheckpointCertificate,
+  Ledger
+}
+import io.iohk.metronome.checkpointing.service.storage.LedgerStorage
+import io.iohk.metronome.crypto.ECPublicKey
+import io.iohk.metronome.hotstuff.consensus.basic.{Phase, QuorumCertificate}
+import io.iohk.metronome.hotstuff.service.ApplicationService
+import io.iohk.metronome.hotstuff.service.storage.{
+  BlockStorage,
+  ViewStateStorage
+}
+import io.iohk.metronome.storage.KVStoreRunner
+
+class CheckpointingService[F[_]: Sync: Applicative, N] private (
+    ledgerTree: Ref[F, Map[Block.Hash, (Ledger, Block.Header)]],
+    lastExecutedHeader: Ref[F, Block.Header],
+    ledgerUpdateSemaphore: Semaphore[F],
+    maxConcurrentUpdates: Int,
+    viewStateStorage: ViewStateStorage[N, CheckpointingAgreement],
+    ledgerStorage: LedgerStorage[N],
+    blockStorage: BlockStorage[N, CheckpointingAgreement]
+)(implicit storeRunner: KVStoreRunner[F, N])
+    extends ApplicationService[F, CheckpointingAgreement] {
+
+  override def createBlock(
+      highQC: QuorumCertificate[CheckpointingAgreement]
+  ): F[Option[Block]] = ???
+
+  override def validateBlock(block: Block): F[Option[Boolean]] = {
+    for {
+      lastExec <- lastExecutedHeader.get
+      ledger <- ledgerUpdateSemaphore.withPermit(
+        updateLedger(block, lastExec.height)
+      )
+      ledgerHashValid = block.header.postStateHash == ledger.hash
+      //TODO: Validate transactions PM-3131/3132
+      txsValid <- Some(false).pure[F]
+    } yield txsValid.map(_ && ledgerHashValid)
+  }
+
+  override def executeBlock(
+      block: Block,
+      commitQC: QuorumCertificate[CheckpointingAgreement],
+      commitPath: NonEmptyList[Block.Hash]
+  ): F[Boolean] = {
+    require(commitQC.phase == Phase.Commit, "Commit QC required")
+
+    if (commitQC.blockHash != block.hash) {
+      lastExecutedHeader.get
+        .flatMap { h =>
+          ledgerUpdateSemaphore.withPermit(
+            updateLedger(block, h.height)
+          )
+        }
+        .as(false)
+    } else {
+      Sync[F]
+        .bracket {
+          ledgerUpdateSemaphore.acquireN(maxConcurrentUpdates)
+        } { _ =>
+          for {
+            lastExec <- lastExecutedHeader.get
+            ledger   <- updateLedger(block, lastExec.height)
+            chkpOpt  <- constructCheckpoint(ledger, commitQC)
+            _        <- commitState(block.header, ledger)
+            -        <- chkpOpt.map(pushCheckpoint).getOrElse(().pure[F])
```

Comment:
This could be outside the lock, since it's not needed to protect the state any more and does some IO.

Reply:
No lock anymore